Bricks/strings/chunk.h

```diff
@@ -0,0 +1,237 @@
+
+  // Returns true and initializes `result` with the rest of this `Chunk`, if it starts with a given prefix.
+  // OK to pass in `result` same as `this`.
+  bool ExpungePrefix(const Chunk& prefix, Chunk& result) const {
+    if (HasPrefix(prefix)) {
+      result.assign(S + prefix.N, N - prefix.N);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  int LexicographicalCompare(const Chunk& rhs) const {
+    const Chunk& lhs = *this;
+    const int d = ::memcmp(S, rhs.S, std::min(lhs.N, rhs.N));
+    if (d) {
+      return d;
+    } else {
+      if (lhs.N < rhs.N) {
+        return -1;
+      } else if (lhs.N > rhs.N) {
+        return +1;
+      } else {
+        return 0;
+      }
+    }
+  }
+
+  struct HashFunction final {
+    // Must actually consider the eternal nature of `Chunk`, not just its pointer. Sigh. Inequality at its best.
+    size_t operator()(const Chunk& chunk) const {
+      // TODO(dkorolev): Use a better hash one day.
+      double hash = 0.0;
+      double k = 1.0;
+      for (size_t i = 0; i < chunk.N; ++i, k = cos(k + i)) {
+        hash += k * chunk.S[i];
+      }
+      return static_cast<size_t>(hash) ^ static_cast<size_t>(hash * 0x100) ^
+             static_cast<size_t>(hash * 0x10000);
+    }
+  };
+
+  struct LexicographicalComparator final {
+    bool operator()(const Chunk& lhs, const Chunk& rhs) const { return lhs.LexicographicalCompare(rhs) < 0; }
+  };
+
+  struct EqualityComparator final {
+    bool operator()(const Chunk& lhs, const Chunk& rhs) const { return lhs.LexicographicalCompare(rhs) == 0; }
+  };
+
+  typedef EqualityComparator Pride;  // Your favorite equality smiley here. Mine is *HAWAII*, because rainbow.
+
+ private:
+  const char* S;
+  size_t N;
+
+  friend class UniqueChunk;
+};
+
+// By [intelligent] design, the length of the string is incorporated in the very pointer to `UniqueChunk`.
+// Even with the same underlying `const char*` pointer, `ChunkDB` will use different `UniqueChunk`-es
+// for `Chunk`-es of different length. Thus `UniqueChunk` doesn't have to consider `Chunk::N` at all.
+class UniqueChunk final : public Chunk {
+ public:
+  UniqueChunk() = default;
+  ~UniqueChunk() = default;
+  UniqueChunk(const UniqueChunk&) = default;
+  UniqueChunk(UniqueChunk&&) = default;
+  UniqueChunk& operator=(const UniqueChunk&) = default;
+  UniqueChunk& operator=(UniqueChunk&&) = default;
+
+  UniqueChunk(const Chunk& chunk) : Chunk(chunk) {}
+  void operator=(const Chunk& chunk) { S = chunk.S, N = chunk.N; }
+
+#ifdef DEFINE_COMPARATOR
+#error "`DEFINE_COMPARATOR()` already defined. No good."
+#endif
+
+#define DEFINE_COMPARATOR(OP) \
+  bool operator OP(const UniqueChunk& rhs) const { return S OP rhs.S; }
+  // It's `clang-format` inserting those spaces between the operator and closing parenthesis; not me. -- D.K.
+  DEFINE_COMPARATOR(== );
+  DEFINE_COMPARATOR(!= );
+  DEFINE_COMPARATOR(< );
```

Comment:
They don't have to compare lexicographically -- just form a total order for `std::map`'s red-black tree to work.